opensearch-project/OpenSearch
# [Tiered Caching] Bug fix for IndicesRequestCache StaleKey management

### Description

Indices Request Cache Stale Key Management introduced in https://github.com/opensearch-project/OpenSearch/pull/12625 had a bug that would have resulted in inaccurate `staleKeysCount`. 

The solution I have come up with is to decrement the staleKeysCount if the readerCacheKeyId has entries in the map, if the map does not have an entry, that means it has been accounted for. 

I have modified the logic to skip a notification with Removal Reason of `REPLACED`, since `REPLACED` is used when a value of cache is updated but not the key. 

All other removal reasons are when a cache entry is deleted and hence is relevant to the staleness calculation. 

##### What was the bug ?
The `incrementStaleKeysCount` method would remove the entries from the cleanupKeyToCountMap if the count of a readerCacheKeyId would be 0 and/or if the shardId had no entry. This way we would avoid the problem of double counting. This method would also increment the staleKeysCount accordingly. 

But on receiving an onRemoval notification on a cache entry being deleted, we would look for the same entry (previously deleted) in the map and deduct the staleKeysCount and hence end up never deducting the staleKeysCount. 

##### Why was this not caught in tests ?
There was a bug [in this UT](https://github.com/opensearch-project/OpenSearch/pull/12625/files#diff-d8997cbc24290f6c5fe02e3bee8f3652874fcc2e3fd4c6fda872cdc869076641R529). We closed the `reader` to create stale keys, but the removal notification we sent was of the `secondReader` and hence we were able to look up the key for in the cleanupKeyToCountMap and hence deduct the staleKeysCount from 1 to 0. 


#### Manual testing for race condition
For the scenario of a concurrent operation like refresh/close/delete closing a reader right after adding an entry into the cache and before registering the close listener to the reader - we were able to manually verify this with 

```java
BytesReference value = cache.computeIfAbsent(key, cacheLoader);
IndexShard indexShard = (IndexShard) cacheEntity.getCacheIdentity();
indexShard.refresh("test");
```

confirmed this would 
- add entry into the map 
- enqueue the cleanupKey after the reader actually closes once the search completes. 
- removes the entry from the map
- increments the staleKeyCount

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/12540


### Check List
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
